internal/runtime/client/client.go

```diff
@@ -74,180 +92,247 @@ func (c *client) Extension(ext *runtimev1.Extension) ExtensionClient {
-}
+type ExtensionClient interface {
+	// Discover makes the discovery call on the extension and updates the runtime extensions
+	// information in the extension status.
+	// TODO: Need a final decision on if we also want to run register inside discover.
+	Discover() (*runtimev1.Extension, error)
 
-type SpecVersion string
+	// Register registers the extension with the manager.
+	Register() error
 
-func (o SpecVersion) ApplyToServiceOptions(opts *ServiceOptions) {
-	opts.Version = string(o)
+	//Unregister unregisters the extension with the manager.
+	Unregister() error
 }
 
-type ServiceOptions struct {
-	Version string // TODO: pointer?
+var _ ExtensionClient = &extensionClient{}
+
+type extensionClient struct {
+	client *client
+	ext    *runtimev1.Extension
 }
 
-type ServiceClient interface {
-	Invoke(ctx context.Context, in, out runtime.Object) error
+func (e *extensionClient) Discover() (*runtimev1.Extension, error) {
+	gvh, err := e.client.catalog.GroupVersionHook(hooksv1alpha1.Discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &hooksv1alpha1.DiscoveryHookRequest{}
+	response := &hooksv1alpha1.DiscoveryHookResponse{}
+
+	// Future work: The discovery runtime extension could be operating on a different hook version than
+	// the latest. We will have to loop through different versions of the discover hook here to actually
+	// finish discovery.
+	opts := &httpCallOptions{
+		catalog: *e.client.catalog,
+		config:  e.ext.Spec.ClientConfig,
+		gvh:     gvh,
+	}
+	if err := httpCall(context.Background(), request, response, opts); err != nil {
+		return nil, errors.Wrap(err, "failed to call the discovery extension")
+	}
+
+	modifiedExtension := &runtimev1.Extension{}
+	e.ext.DeepCopyInto(modifiedExtension)
+	for _, extension := range response.Extensions {
+		modifiedExtension.Status.RuntimeExtensions = append(
+			modifiedExtension.Status.RuntimeExtensions,
+			runtimev1.RuntimeExtension{
+				Name:           extension.Name + "." + e.ext.Name,
+				Hook:           runtimev1.Hook(extension.Hook),
+				TimeoutSeconds: extension.TimeoutSeconds,
+				FailurePolicy:  (*runtimev1.FailurePolicyType)(extension.FailurePolicy),
+			},
+		)
+	}
+
+	// TODO: Decide if we also want to register inside this function.
+
+	return modifiedExtension, nil
 }
 
-type serviceClient struct {
-	client *client
-	svc    catalog.Hook
-	opts   []ServiceOption
+func (e *extensionClient) Register() error {
+	return e.client.registry.Add(e.ext)
 }
 
-var _ ServiceClient = &serviceClient{}
+func (e *extensionClient) Unregister() error {
+	return e.client.registry.Remove(e.ext)
+}
 
-func (c *client) ServiceOld(svc catalog.Hook, opts ...ServiceOption) ServiceClient {
-	return &serviceClient{
-		client: c,
-		svc:    svc,
-		opts:   opts,
-	}
+type httpCallOptions struct {
+	// TODO: Left catalog as not a pointer as the zero value of a pointer type is
+	// not acceptable in the code. The code doest not work if the catalog is nil.
+	catalog       catalog.Catalog
```

Comment:
I think this func won't work if the catalog is not set (and thus just an empty catalog)

I think we should use a `*catalog.Catalog*` instead to avoid copying the entire catalog and its embedded scheme on each call.

We have control over consumers of this func, so it's safe to assume catalog is always set
(if we don't want to assume that we can check and return an error if catalog is nil). But I think we have to avoid that expensive copy.